### DISCUSSION TITLE
libgcc_s_dw2-1.dll

### DISCUSSION THREAD

**Discussion by caobug** [jedisct1/libsodium#1061] [Q&A]
Hello, I use MSYS2+MinGW to compile the DLL file under the Windows platform, but it depends on "libgcc_s_dw2-1.dll".

I added "-static-libgcc -static-libstdc++" to CFLAGS and LDFLAGS, but it didn't work.

![PIC](https://user-images.githubusercontent.com/7737673/118397574-45a1bd80-b687-11eb-888a-e8e6247cd82a.png)

```
#! /bin/sh

NPROCESSORS=$(getconf NPROCESSORS_ONLN 2>/dev/null || getconf _NPROCESSORS_ONLN 2>/dev/null)
PROCESSORS=${NPROCESSORS:-3}

export CFLAGS="-Ofast -fomit-frame-pointer -m32 -march=pentium3 -mtune=westmere -static-libgcc -static-libstdc++"
export LDFLAGS="-static-libgcc -static-libstdc++"
export PREFIX="$(pwd)/libsodium-win32"

if (i686-w64-mingw32-gcc --version > /dev/null 2>&1) then
  echo MinGW found
else
  echo Please install mingw-w64-i686-gcc >&2
  exit
fi

./configure --prefix="$PREFIX" --exec-prefix="$PREFIX" \
            --host=i686-w64-mingw32 && \
make clean && \
make -j${PROCESSORS} && \
make install
```

**Comment by jedisct1** [comment]
If you changed the configuration between two builds, don’t forget to `make distclean`.

**Reply by caobug** [reply]
Yes, I can see the output `checking whether C compiler accepts -Ofast -fomit-frame-pointer -m32 -march=pentium3 -mtune=westmere -static-libgcc -static-libstdc++ -Wl,-Bstatic -lstdc++ -lwinpthread -Wl,-Bdynamic -fvisibility=hidden -fno-strict-aliasing -fno-strict-overflow -fno-asynchronous-unwind-tables -Wall... yes`

but it can't static link this LIBGCC.

**Reply by jedisct1** [reply]
Weird, it seems to be working here (cross-compiled from Linux). And the dependency I see without this is `libgcc_s_seh-1`. No traces of `libgcc_s_dw2-1`.

But is adding `libgcc_s_*` to your project an issue, since you are already adding the libsodium DLL?

Or maybe you can use the `.dll.a` file, which is a static library?

**Reply by caobug** [reply]
https://forum.qt.io/topic/62614/remove-dependency-to-libwinpthread-1-dll/12

Yes, it's really strange. It took 10 hours and still got nothing.

**Reply by ektrah** [reply]
Doesn't the `-Wl,-Bdynamic` undo the previous `-Wl,-Bstatic`?

Does maybe [this](https://stackoverflow.com/a/4986253) or [this](https://stackoverflow.com/a/31037175) help?

**Reply by caobug** [reply]
No, If I add this flag `-static`, it will not compile successfully.

I successfully compiled it a year ago and didn't depends on LIBGCC_*.dll, I have the copy of libsodium-23.dll, but I forgot that compilation process.

**Comment by caobug** [comment]
I tried to use cmake tool to build this project and add `-static -fstack-protector-all` to `cflags`, everything is normal.

Maybe because the libsodium toolchain  doesn't support gcc static flag?

https://github.com/myd7349/libsodium.cmake

```
cmake .. -G 'MinGW Makefiles' -DCMAKE_C_FLAGS="-static -fstack-protector-all" \
         -DBUILD_TESTING=0 -DBUILD_SHARED_LIBS=1
cmake --build . -j${PROCESSORS} --config Release
```

**Reply by jedisct1** [reply]
Regular builds already include both dynamic libraries (in `bin/`) and static libraries (in `lib/`).

**Reply by caobug** [reply]
It didn't complete because it went wrong. you can try to modify `-static-libgcc -static-libstdc++` to `-static` in `CFLAGS` and try again.

```
 /usr/bin/install -c -m 644  sodium.h '/tmp/libsodium-stable/libsodium-win32/include/.'
 /usr/bin/mkdir -p '/tmp/libsodium-stable/libsodium-win32/include'
 /usr/bin/mkdir -p '/tmp/libsodium-stable/libsodium-win32/include/sodium'
 /usr/bin/install -c -m 644  sodium/version.h '/tmp/libsodium-stable/libsodium-win32/include/sodium'
make[4]: Leaving directory '/tmp/libsodium-stable/src/libsodium/include'
make[3]: Leaving directory '/tmp/libsodium-stable/src/libsodium/include'
make[3]: Entering directory '/tmp/libsodium-stable/src/libsodium'
make[4]: Entering directory '/tmp/libsodium-stable/src/libsodium'
 /usr/bin/mkdir -p '/tmp/libsodium-stable/libsodium-win32/bin'
 /usr/bin/install -c -m 644 ./libsodium-24.def '/tmp/libsodium-stable/libsodium-win32/bin'
/usr/bin/install: cannot stat './libsodium-24.def': No such file or directory
make[4]: *** [Makefile:3086: install-defexecDATA] Error 1
make[4]: Leaving directory '/tmp/libsodium-stable/src/libsodium'
make[3]: *** [Makefile:3277: install-am] Error 2
make[3]: Leaving directory '/tmp/libsodium-stable/src/libsodium'
make[2]: *** [Makefile:3113: install-recursive] Error 1
make[2]: Leaving directory '/tmp/libsodium-stable/src/libsodium'
make[1]: *** [Makefile:407: install-recursive] Error 1
make[1]: Leaving directory '/tmp/libsodium-stable/src'
make: *** [Makefile:523: install-recursive] Error 1
```

**Reply by caobug** [reply]
If I  skip build `def` (https://github.com/jedisct1/libsodium/blob/master/src/libsodium/Makefile.am#L193), although it can be compiled successfully, but no `libsodium-*dll` files are generated, only `libsodium.a`
